### PR TITLE
Improve performance of dhconv contraction

### PR DESCRIPTION
**Pull request by mcgibbon**
This PR changes the shape of the weight matrix in SpectralConvS2 from [group, in, out, lat] to [group, lat, out, in] so that the matrix multiplication dimensions [out, in] are the fast dimensions, with [in] being the fastest. Backwards compatibility is included for existing checkpoints.

Changes:
- No public API changes

- [x] Tests added

### CODE REVIEW

**Comment by mcgibbon** [comment]
Commits are done such that there's verification the model produces identical behavior. I made a separate commit updating the regression target which only trivially changes the weight initialization (making it cleaner, but breaking regression).

Still need to write a test that ensures previous checkpoints have their behavior unchanged, but it should be implemented/working using the same method we use to support old checkpoints without a group dim.

Before and after (note the change to dhconv and the total):
image
image
Note that with 8 groups, we no longer see an improvement as the runtime is already low (15ms on T4).

**Comment by mcgibbon** [comment]
~~When reviewing, please check that the regression file updates are valid by looking per-commit. The weight shape refactor is done first in a way that maintains backwards compatibility and passes regression tests, with the regression targets only updated afterwards with a trivial change to the way the weights are initialized (which changes the values they get initialized to).~~

I added a checkpoint-load regression test in another PR, so this is no longer an issue.

**Comment by mcgibbon** [comment]
I added a regression test for the output from a loaded checkpoint being the same, using an input checkpoint generated from main. The test caught a bug, so it's working properly.

File: fme/core/models/conditional_sfno/s2convolutions.py [hunk 1]
```diff
@@ -59,7 +59,9 @@ def _contract_lora(
     """
     lora_A = torch.view_as_complex(lora_A)
     lora_B = torch.view_as_complex(lora_B)
-    return torch.einsum("girx,grox,bgixy->bgoxy", lora_A, lora_B, x)
+    # tmp = torch.einsum("gxri,bgixy->bgxry", lora_A, x)
```
Comment (Arcomano1234) [inline fme/core/models/conditional_sfno/s2convolutions.py]: I believe this is left over commented out code from testing

File: fme/core/models/conditional_sfno/s2convolutions.py [hunk 2]
```diff
@@ -179,38 +181,40 @@ def __init__(
...
             self.modes_lat_local,
+            out_channels // num_groups,
+            in_channels // num_groups,
         ]
 
         assert factorization == "ComplexDense"
         self.weight = nn.Parameter(scale * torch.randn(*weight_shape, 2))
-        self.weight.is_shared_mp = ["matmul", "w"]
 
+        self.lora_rank = lora_rank
```
Comment (Arcomano1234) [inline fme/core/models/conditional_sfno/s2convolutions.py:198]: `self.lora_rank `looks unused

File: fme/core/models/conditional_sfno/s2convolutions.py [hunk 3]
```diff
@@ -228,6 +232,68 @@ def __init__(
 
         # rewrite old checkpoints on load
         self.register_load_state_dict_pre_hook(self._add_singleton_group_dim)
+        self.register_load_state_dict_pre_hook(self._reorder_weight_dims)
+
+    @staticmethod
+    def _reorder_weight_dims(
```
Comment (Arcomano1234) [inline fme/core/models/conditional_sfno/s2convolutions.py:236]: Optional: Its a little hard for me to say for sure this is working 100% correctly. Did we want to add a test with the old ckpt order to make sure this working? If you have already tested this then its fine but I just want to make sure there isn't any subtle bugs like weight naming bugs etc.

Comment (mcgibbon) [reply fme/core/models/conditional_sfno/s2convolutions.py:236]: I added (possibly in an earlier pr) a regression test that asserts if you load a presaved checkpoint and call it, the result should be the same as the presaved target. It’s maybe worth convincing yourself of this for the review.

Comment (Arcomano1234) [reply fme/core/models/conditional_sfno/s2convolutions.py:236]: I see, yes this was in an early PR. Seems like that test passes so `_reorder_weight_dims` must be doing its job correctly.

File: fme/core/models/conditional_sfno/s2convolutions.py (hunk 2 above)
Comment (mcgibbon) [reply fme/core/models/conditional_sfno/s2convolutions.py:198]: It's used as `module.lora_rank` in `_reorder_weight_dims`

File: fme/core/models/conditional_sfno/s2convolutions.py (hunk 1 above)
Comment (mcgibbon) [reply fme/core/models/conditional_sfno/s2convolutions.py]: Sort of, it's an important performance optimization I'll be merging in the next PR. Still, I suppose I can remove it.

Comment (Arcomano1234) [reply fme/core/models/conditional_sfno/s2convolutions.py]: I see in that case it doesn't matter I just didnt know if this was intentional

File: fme/core/models/conditional_sfno/s2convolutions.py (hunk 2 above)
Comment (Arcomano1234) [reply fme/core/models/conditional_sfno/s2convolutions.py:198]: Ahh I forgot about that
